internal/datanode/data_sync_service.go

```diff
@@ -142,15 +145,52 @@ func (dsService *dataSyncService) close() {
 	dsService.flushManager.close()
 }
 
+// getSegmentInfos return the SegmentInfo details according to the given ids through RPC to datacoord
+func (dsService *dataSyncService) getSegmentInfos(segmentIds []int64) ([]*datapb.SegmentInfo, error) {
+	var segmentInfos []*datapb.SegmentInfo
+	err := retry.Do(context.Background(), func() error {
+		infoResp, err := dsService.dataCoord.GetSegmentInfo(dsService.ctx, &datapb.GetSegmentInfoRequest{
+			Base: &commonpb.MsgBase{
+				MsgType:   commonpb.MsgType_SegmentInfo,
+				MsgID:     0,
+				Timestamp: 0,
+				SourceID:  Params.ProxyCfg.GetNodeID(),
+			},
+			SegmentIDs: segmentIds,
+		})
+		if err != nil {
+			log.Error("Fail to get datapb.SegmentInfo by ids from datacoord", zap.Error(err))
+			return err
+		}
+		if infoResp.GetStatus().ErrorCode != commonpb.ErrorCode_Success {
+			err = errors.New(infoResp.GetStatus().Reason)
+			log.Error("Fail to get datapb.SegmentInfo by ids from datacoord", zap.Error(err))
+			return err
+		}
+		segmentInfos = infoResp.Infos
+		return nil
+	}, retry.Attempts(10), retry.Sleep(100*time.Millisecond))
+	if err != nil {
+		log.Error("Fail to get datapb.SegmentInfo by ids from datacoord", zap.Error(err))
+		return nil, err
+	}
+	return segmentInfos, nil
+}
+
 // initNodes inits a TimetickedFlowGraph
 func (dsService *dataSyncService) initNodes(vchanInfo *datapb.VchannelInfo) error {
 	dsService.fg = flowgraph.NewTimeTickedFlowGraph(dsService.ctx)
 	// initialize flush manager for DataSync Service
 	dsService.flushManager = NewRendezvousFlushManager(dsService.idAllocator, dsService.chunkManager, dsService.replica,
 		flushNotifyFunc(dsService), dropVirtualChannelFunc(dsService))
 
+	var err error
 	// recover segment checkpoints
-	for _, us := range vchanInfo.GetUnflushedSegments() {
+	unflushedSegmentInfos, err := dsService.getSegmentInfos(vchanInfo.GetUnflushedSegmentIds())
```

Comment:
There might be still some room to clean the cleanup the code, but it's ok for now